### ISSUE TITLE
Improve "Getting Started" experience by disabling NVD mirroring per default

### ISSUE THREAD

**Issue by nscuro** [DependencyTrack/dependency-track#2524]
### Current Behavior

Dependency-Track automatically kicks off the NVD mirroring process just one minute after it starts.

Mirroring the NVD is an expensive operation that causes large amounts of memory allocations. With the current mirroring implementation, heap size grows from ~250mb to >3gb (with `-Xmx8g`) in a very short time frame. Because we use `-XX:MaxRAMPercentage=90.0` in our Dockerfile, the JVM has a low incentive of GCing as much. Clearly we have room for improvement here, but by nature of the task itself, it will always be heavyweight.

I have heard on multiple occasions now that this behavior makes it very hard to get started, or test custom integrations locally. Spinning up DT for integration tests via [testcontainers](https://www.testcontainers.org/) is almost out of question. It is unfortunate that the first impression of DT some people have is "Wow this thing needs a lot of RAM".

While the NVD data is "nice to have", I think it's fair to say that the majority of CVEs hosted in the NVD will be irrelevant to most users. This is further amplified by the fact that most SBOMs to not contain CPEs, which means that DT will not be able to match against NVD data anyway. The NVD is the only data source enabled by default, and it is most likely the least used, due to not including PURLs.

Idealy, DT should start quickly, and not kick off any heavyweight tasks unless explicitly asked to.

At the very least, it will make DT more accessible by allowing users with "underpowered" workstations to kick the tires more easily.

### Proposed Behavior

Disable NVD mirroring by default, and require explicit opt-in to enable it. This behavior will be analog to how it's done for GHSA, OSV, and VulnDB. For existing installations, we can assume it's enabled.

Given the fact that the NVD will require usage of API keys soon (#1861), users will have to provide additional details anyway before mirroring becomes viable (it will be way too slow without API key).

Separately, revisiting potential improvements in the NVD mirroring process is also something we should do.

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this enhancement was already requested

**Comment by stevespringett** [comment]
Please do not implement this without a default fall-back behavior which allows users to discover if they are impacted by a new popular vulnerability. 

The entire reason NTIA met and why SBOM became a thing is to discover what is affected, and where.  Our docs have a section related to this. https://docs.dependencytrack.org/usage/impact-analysis/

This use case is the reason why DT is a reference implementation for SBOM in the first place.

We need a fall-back plan. Disabling NVD mirroring without the ability to accomplish this use case is a non-starter.

**Comment by nscuro** [comment]
@stevespringett I am actually looking into lowering the footprint of the mirroring task first. That undoubtedly has a higher benefit over the proposal I brought forward here.

For my understanding: Wouldn't that use-case still be achievable by simply enabling NVD mirroring? Is having it enabled by default a hard requirement for this use-case?

**Comment by stevespringett** [comment]
> Wouldn't that use-case still be achievable by simply enabling NVD mirroring? 

Yes, but assuming a user knows that these two things are related is a poor assumption. The out of the box behavior needs to support this primary use case. If NVD mirroring is disabled by default, what is the proposal to support this use case without NVD data present?

One possible way would be to mirror on demand. Another would be to ship a lucene index already prepopulated that we could fall back to. There are other fallback mechanism as well. But this primary use case needs to be supported out of the box.

**Comment by stevespringett** [comment]
Also, I'm not sure if vulnerability aliasing would have issues if the NVD mirroring is disabled and the user never enabled it.

**Comment by valentijnscholten** [comment]
Is a full mirror needed to support impact analysis out of the box? Without a mirror, analysis can still take place online via the NVD api? For casual users trying out DT it might be beneficial get started quickly without having wait 20-30 minutes to mirror 250k vulnerabilities from the 90s ;-) 

The APIv2 might make the initial mirroring faster. The [docs ](https://nvd.nist.gov/developers/start-here) seem to suggest that, without an apikey, you can do ~10 requests per minute. Each request has a limit of max 2000 results. So the initial download would roughly take 250k/2000/10 = 12,5 minutes. It willl take longer if you run into any ratelimiting causing exponential backoff or whatever other delays.

**Comment by stevespringett** [comment]
> Is a full mirror needed to support impact analysis out of the box?

Mirroring starts with the current year and goes back in time until the beginning. There's no way of predicting what CVE a user will search for when evaluating the impact analysis capabilities. Typeahead could potentially be used for mirroring on demand use cases.

**Comment by nscuro** [comment]
> The APIv2 might make the initial mirroring faster.

The download itself was never a problem, as the [data feeds](https://nvd.nist.gov/vuln/data-feeds) we consume currently are just ~5mb each compressed, ~100mb uncompressed each. Download and decompression of all feeds takes just a few minutes, if even that. It's the processing of the data that is eating too much time and resources.

But IIRC the new API will allow us to only fetch what changed since we last initiated a mirroring operation, meaning we don't have to re-download and re-process thousands of CVEs, just because one of them changed. The initial mirroring will remain heavy lifting though.

BTW, I was able to reduce the heap utilization of the NVD mirroring down to just a few hundred megabytes :)
It also finished ~10min earlier than most of my previous test runs, but not sure how reliable that measurement is.
﻿
<img width="602" alt="image" src="https://user-images.githubusercontent.com/5693141/222020594-d3bc78e1-a0c1-43a7-bb36-a15346731320.png">

**Comment by valentijnscholten** [comment]
> > Is a full mirror needed to support impact analysis out of the box?
> 
> Mirroring starts with the current year and goes back in time until the beginning. There's no way of predicting what CVE a user will search for when evaluating the impact analysis capabilities. Typeahead could potentially be used for mirroring on demand use cases.


It might be wrong to assume users understand that after the first year has been imported, they can start uploading BOMs.
My point was more that NVD also has an online analysis capability. Having all vulnerabilities locally is probably what you want in the end, but if you want to be able to have a lightweight start the online analysis could be something to look at.

**Comment by valentijnscholten** [comment]
> > The APIv2 might make the initial mirroring faster.
> 
> The download itself was never a problem, as the [data feeds](https://nvd.nist.gov/vuln/data-feeds) we consume currently are just ~5mb each compressed, ~100mb uncompressed each. Download and decompression of all feeds takes just a few minutes, if even that. It's the processing of the data that is eating too much time and resources.


I was assuming (bad me!) that downloads and processing can be done in parallel. With only one request every six seconds there's plenty of room to do some processing on the side.

**Comment by roadSurfer** [comment]
Being able to optionally disable the initial mirror would be a great help for dev/testing of new features/fixes.
I do agree though, it needs to be enabled by default otherwise people could have a false sense of security.

**Comment by robertlagrant** [comment]
If this is primarily a testing feature, a possible implementation could be to have the toggle as an environment variable and/or config value. Many settings there (e.g. LDAP query) need to be handled carefully, so it might fit well.

**Comment by wkoot** [comment]
As of 20 August 2025, the feeds that are automatically synced on startup have been decommissioned https://www.nist.gov/itl/nvd#Aug2025

**Comment by nscuro** [comment]
@wkoot Hence why we moved to the new 2.0 data feeds :) https://docs.dependencytrack.org/changelog/#v4-13-4

**Comment by yodatak** [comment]
And adding a way to add env variable to download from alternate url could be good to have because it seem the app is broken on  off grid network
